File: max_heapify/C/max_heapify.c
#include<stdio.h>
void swap(int* A, int a, int b) {
  int tmp = A[a];
  A[a] = A[b];
  A[b] = tmp;
}

void max_heapify(int* A, int m, int i) {
  int left = 2 * i + 1;
  int right = 2 * i + 2;
  int greater = i;

  if (left < m && A[left] > A[i]) {
    greater = left;
  }

  if (right < m && A[right] > A[greater]) {
    greater = right;
  }

  if (greater != i) {
    swap(A, greater, i);
    max_heapify(A, m, greater);
  }
}

int main() {
    int A[9] = { 8, 18, 14, 17, 12, 13, 11, 15, 16 };
    int m = 9;
    int i = 0;

    max_heapify(A, m, i);

    for (i = 0; i < m; i++)
        printf ("%d ", A[i]);

    return 0;
}

// expected: 18 17 14 16 12 13 11 15 8

File: heap/ruby/max_heapify.rb
def max_heapify(heap, wrong_element)
  left = 2 * wrong_element + 1
  right = 2 * wrong_element + 2

  greater = (left <= heap.length && heap[left] > heap[wrong_element]) ? left : wrong_element
  greater = right if right <= heap.length && heap[right] > heap[greater]

  return if greater == wrong_element

  tmp = heap[wrong_element]
  heap[wrong_element] = heap[greater]
  heap[greater] = tmp
  max_heapify(heap, greater)
end

heap = [8, 18, 14, 17, 12, 13, 11, 15, 16]

max_heapify(heap, 0)

puts heap

# expected: 18 17 14 16 12 13 11 15 8
